In [171]:
import numpy as np
import pandas as pd
import seaborn as sb
import re
import src.cleaning_functions as cf

In [172]:
!pwd

/mnt/c/Users/beasp/IronHack/PROYECTOS/proyecto1_pandas


In [173]:
df_attacks = pd.read_csv("data/attacks.csv",encoding = "ISO-8859-1") #To load the data csv
#mi dataFrame es df_attacks

Hipóteis:
- Se dan más ataques a hombres que practican surf que en hombres que practican natación
- Se dan más ataques en junio que en febrero en California
- El tiburón White shark es el que provoca más ataques en USA

1.Ahora voy a eliminar las columnas que no me van a dar ninguna información para comprobar mis hipótesis, en principio las que voy a usar SEGURO son:
- Sex
- Activity
- Date
- Area
- Species
- Country

In [174]:
df_attacks.head(5)

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [175]:
borrar= df_attacks.drop(["Case Number","pdf","href formula","href","Case Number.1","Case Number.2","original order","Unnamed: 22","Unnamed: 23"], axis=1)
#meto en variable "borrar" para borrar las columnas que sé que no me van a hacer falta para mis hipótesis

In [6]:
borrar.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  
0          18h00      White shark                R. Collier, GSAF  
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com  
2          07h45              NaN  K.McMurray, TrackingSharks.com  
3            NaN        2 m shark                  B. Myatt, GSAF  
4            NaN  Tiger shark, 3m                       A .Kipper

In [176]:
df_attacks.shape

(25723, 24)

2.Voy a eliminar las filas duplicadas y filas con valores vacíos

In [177]:
borrar_duplicados= borrar.drop_duplicates() #me creo una variable para poder seguir trabajando con este dato, como por ejemplo para saber su shape a continuación
borrar_duplicados.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  
0          18h00      White shark                R. Collier, GSAF  
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com  
2          07h45              NaN  K.McMurray, TrackingSharks.com  
3            NaN        2 m shark                  B. Myatt, GSAF  
4            NaN  Tiger shark, 3m                       A .Kipper

In [179]:
borrar_duplicados.dropna(how="all",inplace=True) #solo había una fila coon todos los valores vacíos
borrar_duplicados.shape

(6301, 15)

Hemos pasado de tener (25723 filas, 24) a tener (6312 filas, 24)

3.Ahora voy a reemplazar los nan por Unknown/Invalid para cada una de las columnas, pasando a ser un string. Para ello reviso primero cuáles de los valores son categóricos.

In [180]:
borrar_duplicados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6301 entries, 0 to 6301
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Date                    6301 non-null   object 
 1   Year                    6299 non-null   float64
 2   Type                    6297 non-null   object 
 3   Country                 6251 non-null   object 
 4   Area                    5846 non-null   object 
 5   Location                5761 non-null   object 
 6   Activity                5757 non-null   object 
 7   Name                    6091 non-null   object 
 8   Sex                     5736 non-null   object 
 9   Age                     3471 non-null   object 
 10  Injury                  6273 non-null   object 
 11  Fatal (Y/N)             5762 non-null   object 
 12  Time                    2948 non-null   object 
 13  Species                 3463 non-null   object 
 14  Investigator or Source  6284 non-null   

In [182]:
borrar_duplicados["Date"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Year"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Type"].fillna(("Invalid"),inplace=True)
borrar_duplicados["Country"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Area"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Location"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Activity"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Name"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Sex "].fillna(("Unknown"),inplace=True)
borrar_duplicados["Age"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Injury"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Fatal (Y/N)"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Time"].fillna(("Unknown"),inplace=True)
borrar_duplicados["Species "].fillna(("Unknown"),inplace=True)
borrar_duplicados["Investigator or Source"].fillna(("Unknown"), inplace=True)

In [183]:
borrar_duplicados.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex       Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F       57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F       11   
2                    Habush, Oahu      Surfing      John Denges    M       48   
3              Arrawarra Headland      Surfing             male    M  Unknown   
4                        La Ticla  Free diving   Gustavo Ramos     M  Unknown   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  
0          18h00      White shark                R. Collier, GSAF  
1  14h00  -15h00          Unknown  K.McMurray, TrackingSharks.com  
2          07h45          Unknown  K.McMurray, TrackingSharks.com  
3        Unknown        2 m shark                  B. Myatt, GSAF  
4        Unknown  Tiger shark, 3m                       A .Kipper

4.Comienzo con la sustitución de valores dentro de las columnas que me van a dar información para comprobar mis hipótesis.

4.1 COLUMNA SEX: voy a ver los valores únicos de la columna Sex, que al ser valores categóricos realmente solo me van a dar información si aparece uno de los dos géneros, o desconocido.

In [184]:
sex= borrar_duplicados["Sex "].unique()
sex

array(['F', 'M', 'Unknown', 'M ', 'lli', 'N', '.'], dtype=object)

Voy a reemplazar los valores que parece que quieren decir lo mismo para que se llamen de una misma manera, y los que no me dicen nada por "Unknown"

In [185]:
reemplazar1=borrar_duplicados.replace({"M ": "M", "lli":"Unknown",".":"Unknown","N":"Unknown"}) 
reemplazar1.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex       Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F       57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F       11   
2                    Habush, Oahu      Surfing      John Denges    M       48   
3              Arrawarra Headland      Surfing             male    M  Unknown   
4                        La Ticla  Free diving   Gustavo Ramos     M  Unknown   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...     Unknown   
1                         Minor injury to left thigh     Unknown   
2       Injury to left lower leg from surfboard skeg     Unknown   
3                          Minor injury to lower leg     Unknown   
4  Lacerations to leg & hand shark PROVOKED INCIDENT     Unknown   

            Time         Species           Investigator or Source  
0          18h00      White shark                R. Collier, GSAF  
1  14h00  -15h00          Unknown  K.McMurray, TrackingSharks.com  
2          07h45          Unknown  K.McMurray, TrackingSharks.com  
3        Unknown        2 m shark                  B. Myatt, GSAF  
4        Unknown  Tiger shark, 3m                       A .Kipper

In [186]:
reemplazar1["Sex "].unique() #compruebo si se han reemplazado los valores indicados, viendo de nuevo los valores únicos de la columna Sex.


array(['F', 'M', 'Unknown'], dtype=object)

4.2 COLUMNA ACTIVITY: a partir de mi df ""reemplazar1" que ya cuenta con la limpieza de una columna, voy a limpiar la columna Activity.

In [187]:
len(reemplazar1["Activity"].unique())

1531

In [188]:
Actividad=reemplazar1.Activity.unique() #hay que hacerlo lista porque sino no salen todas las actividades.
list(Actividad)


['Paddling',
 'Standing',
 'Surfing',
 'Free diving',
 'Kite surfing',
 'Swimming',
 'Fishing',
 'Walking',
 'Feeding sharks',
 'Boogie boarding',
 'Scuba diving',
 'Paddle-skiing',
 'Body boarding',
 'Windsurfing',
 'Stand-Up Paddleboarding',
 'Wading',
 'Scuba Diving',
 'Kayak fishing for sharks',
 'Cleaning fish',
 'Snorkeling',
 'Spearfishing',
 'Unknown',
 'Diving',
 'Fishing / Wading',
 '2 boats capsized',
 'Night bathing',
 'Surfing ',
 'Kayaking / Fishing',
 'Kayaking',
 'Body surfing',
 'Swimming, poaching abalone',
 'Canoeing',
 'SUP',
 'Skimboarding',
 'Touching a shark',
 'Fishing ',
 'Attempting to lasso a shark',
 'Photo shoot',
 'Paddle boarding',
 'Kakaying',
 'Washing hands',
 'Grabbing shark for a selfie',
 'Kayak fishing',
 'Tagging sharks',
 'Surf skiing ',
 'Surf fishing',
 'Floating',
 'Cage Diving',
 'Surfng',
 'SUP Foil boarding',
 'Lobstering',
 'Fishing in Alabama Deep Fishing Rodeo',
 'Fishing for squid',
 'Fishing for sharks',
 'Floating in tube',
 'Teasing 

In [189]:
lista_actividad=str(list(Actividad))
lista_actividad

'[\'Paddling\', \'Standing\', \'Surfing\', \'Free diving\', \'Kite surfing\', \'Swimming\', \'Fishing\', \'Walking\', \'Feeding sharks\', \'Boogie boarding\', \'Scuba diving\', \'Paddle-skiing\', \'Body boarding\', \'Windsurfing\', \'Stand-Up Paddleboarding\', \'Wading\', \'Scuba Diving\', \'Kayak fishing for sharks\', \'Cleaning fish\', \'Snorkeling\', \'Spearfishing\', \'Unknown\', \'Diving\', \'Fishing / Wading\', \'2 boats capsized\', \'Night bathing\', \'Surfing \', \'Kayaking / Fishing\', \'Kayaking\', \'Body surfing\', \'Swimming, poaching abalone\', \'Canoeing\', \'SUP\', \'Skimboarding\', \'Touching a shark\', \'Fishing \', \'Attempting to lasso a shark\', \'Photo shoot\', \'Paddle boarding\', \'Kakaying\', \'Washing hands\', \'Grabbing shark for a selfie\', \'Kayak fishing\', \'Tagging sharks\', \'Surf skiing \', \'Surf fishing\', \'Floating\', \'Cage Diving\', \'Surfng\', \'SUP Foil boarding\', \'Lobstering\', \'Fishing in Alabama Deep Fishing Rodeo\', \'Fishing for squid\',

In [190]:
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Ff](ish|ISH).*","Fishing",regex=True)
reemplazar1.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex       Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F       57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F       11   
2                    Habush, Oahu      Surfing      John Denges    M       48   
3              Arrawarra Headland      Surfing             male    M  Unknown   
4                        La Ticla  Free diving   Gustavo Ramos     M  Unknown   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...     Unknown   
1                         Minor injury to left thigh     Unknown   
2       Injury to left lower leg from surfboard skeg     Unknown   
3                          Minor injury to lower leg     Unknown   
4  Lacerations to leg & hand shark PROVOKED INCIDENT     Unknown   

            Time         Species           Investigator or Source  
0          18h00      White shark                R. Collier, GSAF  
1  14h00  -15h00          Unknown  K.McMurray, TrackingSharks.com  
2          07h45          Unknown  K.McMurray, TrackingSharks.com  
3        Unknown        2 m shark                  B. Myatt, GSAF  
4        Unknown  Tiger shark, 3m                       A .Kipper

Compruebo si solo hay un valor de "fishing"

In [191]:
Actividad=reemplazar1.Activity.unique() 
list(Actividad)

['Paddling',
 'Standing',
 'Surfing',
 'Free diving',
 'Kite surfing',
 'Swimming',
 'Fishing',
 'Walking',
 'Feeding sharks',
 'Boogie boarding',
 'Scuba diving',
 'Paddle-skiing',
 'Body boarding',
 'Windsurfing',
 'Stand-Up Paddleboarding',
 'Wading',
 'Scuba Diving',
 'Snorkeling',
 'Unknown',
 'Diving',
 '2 boats capsized',
 'Night bathing',
 'Surfing ',
 'Kayaking',
 'Body surfing',
 'Swimming, poaching abalone',
 'Canoeing',
 'SUP',
 'Skimboarding',
 'Touching a shark',
 'Attempting to lasso a shark',
 'Photo shoot',
 'Paddle boarding',
 'Kakaying',
 'Washing hands',
 'Grabbing shark for a selfie',
 'Tagging sharks',
 'Surf skiing ',
 'Floating',
 'Cage Diving',
 'Surfng',
 'SUP Foil boarding',
 'Lobstering',
 'Floating in tube',
 'Teasing a shark',
 'Surf-skiing',
 'Diving for beche-de-mer',
 'Feeding stingrays?',
 'Sea disaster',
 'Swimming / Wading',
 'Scallop diving on hookah',
 'Treading water',
 'Body boarding ',
 'Playing in the water',
 'Body Boarding',
 'Body surfing?',

Efectivamente, solo hay un valor de Fishing.

In [192]:
len(Actividad)

1205

Continúo unificando tipos de actividad, renombrandolos para que no haya tantos valores diferentes que se refieran mismo tipo de actividad.

In [193]:
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Ss](wim|WIM).*","Swimming",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Kk](ite|ITE).*","Kite",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Ww](alking|ALKING).*","Walking",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Bb](oog|OOG).*","Boogie Board",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Bb](ody|ODY).*","Body Boarding",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[wW](ind|IND).*","Wind Surfing",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Bb](oat|OAT).*","Boat",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Ss](hark|HARK).*","Interact with sharks",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Dd](iving|IVING).*","Diving",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Ss](tand|TAND).*","Standing in water",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Pp](addl|ADDL).*","Paddling",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Bb](athing|ATHING).*","Bathing",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Oo](verb|VERB).*","OverBoard",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Bb](athing|ATHING).*","Bathing",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Ff](loat|LOAT).*","Floating",regex=True)
reemplazar1["Activity"]= reemplazar1["Activity"].str.replace(".*[Jj](ump|UMP).*","Jumping",regex=True)

In [194]:
len(Actividad)

1205

In [195]:
Actividad_limpio=list(reemplazar1["Activity"].unique())
Actividad_limpio

['Paddling',
 'Standing in water',
 'Surfing',
 'Diving',
 'Kite',
 'Swimming',
 'Fishing',
 'Walking',
 'Interact with sharks',
 'Boogie Board',
 'Body Boarding',
 'Wind Surfing',
 'Wading',
 'Snorkeling',
 'Unknown',
 'Boat',
 'Bathing',
 'Surfing ',
 'Kayaking',
 'Canoeing',
 'SUP',
 'Skimboarding',
 'Photo shoot',
 'Kakaying',
 'Washing hands',
 'Surf skiing ',
 'Floating',
 'Surfng',
 'SUP Foil boarding',
 'Lobstering',
 'Surf-skiing',
 'Feeding stingrays?',
 'Sea disaster',
 'Treading water',
 'Playing in the water',
 'Playing',
 'Sitting in the water',
 'Transatlantic Rowing',
 'Rowing',
 'Surfing & filming dolphins',
 'Filming',
 'Kayaking ',
 'Playing with an air mattress',
 'Kneeling in the water',
 'Fell into the water',
 'Sailing',
 'Filming a documentary',
 'Casting a net',
 'Playing in the surf',
 'Jumping',
 'Kayaking or Wave skiing',
 'Jet skiing',
 'Wakeboarding',
 'Attempting to fix motor',
 'Rescuing',
 'Washing his feet',
 'Crawling',
 'Air Disaster',
 'Crabbing',
 

In [196]:
len(Actividad_limpio)

478

In [197]:
type(Actividad_limpio)

list

In [198]:
frecuencia_actividades= list(reemplazar1.Activity.value_counts(sort=False))
frecuencia_actividades

[37,
 140,
 971,
 501,
 22,
 1109,
 1224,
 26,
 149,
 51,
 136,
 23,
 149,
 89,
 550,
 103,
 185,
 5,
 33,
 13,
 3,
 1,
 1,
 1,
 1,
 7,
 44,
 1,
 1,
 1,
 12,
 1,
 8,
 32,
 2,
 7,
 2,
 2,
 12,
 1,
 1,
 2,
 1,
 1,
 9,
 9,
 1,
 1,
 3,
 15,
 1,
 2,
 1,
 1,
 1,
 2,
 1,
 2,
 3,
 1,
 1,
 1,
 6,
 1,
 1,
 6,
 1,
 1,
 1,
 4,
 3,
 1,
 1,
 1,
 1,
 1,
 12,
 1,
 9,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 7,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 4,
 1,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 3,
 6,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 4,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 5,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 84,
 1,
 1,
 1,
 1,
 3,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 2,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 

In [199]:
len(frecuencia_actividades) 

478

Como se puede observar la longitug de la lista de valores únicos y de frecuencias coincide.

Voy a crear un diccionario donde aparezcan cada actividad con su frecuencia

In [200]:
dicc_actividad ={}
for Actividad_limpio, frecuencia_actividades in zip(Actividad_limpio, frecuencia_actividades):
    dicc_actividad[Actividad_limpio] = frecuencia_actividades

In [222]:
dicc_actividad

{'Paddling': 37,
 'Standing in water': 140,
 'Surfing': 971,
 'Diving': 501,
 'Kite': 22,
 'Swimming': 1109,
 'Fishing': 1224,
 'Walking': 26,
 'Interact with sharks': 149,
 'Boogie Board': 51,
 'Body Boarding': 136,
 'Wind Surfing': 23,
 'Wading': 149,
 'Snorkeling': 89,
 'Unknown': 550,
 'Boat': 103,
 'Bathing': 185,
 'Surfing ': 5,
 'Kayaking': 33,
 'Canoeing': 13,
 'SUP': 3,
 'Skimboarding': 1,
 'Photo shoot': 1,
 'Kakaying': 1,
 'Washing hands': 1,
 'Surf skiing ': 7,
 'Floating': 44,
 'Surfng': 1,
 'SUP Foil boarding': 1,
 'Lobstering': 1,
 'Surf-skiing': 12,
 'Feeding stingrays?': 1,
 'Sea disaster': 8,
 'Treading water': 32,
 'Playing in the water': 2,
 'Playing': 7,
 'Sitting in the water': 2,
 'Transatlantic Rowing': 2,
 'Rowing': 12,
 'Surfing & filming dolphins': 1,
 'Filming': 1,
 'Kayaking ': 2,
 'Playing with an air mattress': 1,
 'Kneeling in the water': 1,
 'Fell into the water': 9,
 'Sailing': 9,
 'Filming a documentary': 1,
 'Casting a net': 1,
 'Playing in the surf'

Quiero trabajar solo con las keys cuyo valor sea 1 para después renombrarlas

In [218]:
actividad_other= dict(filter(lambda x: x[1] ==1, dicc_actividad.items()))
actividad_other

{'Skimboarding': 1,
 'Photo shoot': 1,
 'Kakaying': 1,
 'Washing hands': 1,
 'Surfng': 1,
 'SUP Foil boarding': 1,
 'Lobstering': 1,
 'Feeding stingrays?': 1,
 'Surfing & filming dolphins': 1,
 'Filming': 1,
 'Playing with an air mattress': 1,
 'Kneeling in the water': 1,
 'Filming a documentary': 1,
 'Casting a net': 1,
 'Kayaking or Wave skiing': 1,
 'Wakeboarding': 1,
 'Attempting to fix motor': 1,
 'Rescuing': 1,
 'Crawling': 1,
 'Yacht race': 1,
 'Sinking of the cargo ship Mark Jason': 1,
 'Rowing an inflatable dinghy': 1,
 'The 426-ton cargo ship Mia, laden with cement, capsized in heavy seas ': 1,
 'Wading?': 1,
 'Playing on a sandbar': 1,
 'Shrimping': 1,
 'Playing soccer in the water': 1,
 'Night Surfing': 1,
 'Treading water/ Surfing': 1,
 "Crouching in 2' of water": 1,
 'Swept out to sea by the tsunami, she clung  to a log for 24 hours': 1,
 'Tandem surfing': 1,
 'Air disaster. Flash Airlines Boeing 737 crashed into the Red Sea': 1,
 'Attempting to retreive a dinghy': 1,
 'S

In [219]:
type(actividad_other)

dict

lo transformo en lista

In [234]:
lista_actividad_other=list(actividad_other.keys())
lista_actividad_other

['Skimboarding',
 'Photo shoot',
 'Kakaying',
 'Washing hands',
 'Surfng',
 'SUP Foil boarding',
 'Lobstering',
 'Feeding stingrays?',
 'Surfing & filming dolphins',
 'Filming',
 'Playing with an air mattress',
 'Kneeling in the water',
 'Filming a documentary',
 'Casting a net',
 'Kayaking or Wave skiing',
 'Wakeboarding',
 'Attempting to fix motor',
 'Rescuing',
 'Crawling',
 'Yacht race',
 'Sinking of the cargo ship Mark Jason',
 'Rowing an inflatable dinghy',
 'The 426-ton cargo ship Mia, laden with cement, capsized in heavy seas ',
 'Wading?',
 'Playing on a sandbar',
 'Shrimping',
 'Playing soccer in the water',
 'Night Surfing',
 'Treading water/ Surfing',
 "Crouching in 2' of water",
 'Swept out to sea by the tsunami, she clung  to a log for 24 hours',
 'Tandem surfing',
 'Air disaster. Flash Airlines Boeing 737 crashed into the Red Sea',
 'Attempting to retreive a dinghy',
 'Snorkeling (filming the sardine run)',
 'Playing in the surf with his 2 dogs',
 'Collecting beche-de-me

In [235]:
type(lista_actividad_other)

list

In [236]:
len(lista_actividad_other)

402

Quiero conseguir que para todos los valores de mi lista_actividad_unknown(lista con valores de frecuencia 1) que estén en mi lista Actividad_limpio (lista con todos los valores únicos de la columna Activity), me devuelva el valor "Unknown"

Utilizo un función que he realizado en visual y por tanto tengo que importarme, para conseguir esto, donde "s" va a ser mi lista con valores de frecuencia 1 y "b" va a ser mi lista de completa de valores únicos.

In [209]:
reemplazar1["Activity_2"] = reemplazar1["Activity"].apply(cf.sustitucion) #no se me importa la función desde visual

AttributeError: module 'src.cleaning_functions' has no attribute 'sustitucion'

In [225]:
def sustitucion(s,b):
    for i,p in enumerate(s):
        if p in b:
            b[i]="other"
    return b

In [238]:
sustitucion(lista_actividad_other, Actividad_limpio)

TypeError: 'str' object does not support item assignment

In [239]:
reemplazar1["Actividad"] = reemplazar1["Activity"].apply(sustitucion[1])

TypeError: 'function' object is not subscriptable

In [156]:
#me entiende la función al revés, me sustituye unknown en todos los valores que no forman parte de lista_actividad_unknown...???????????????????????

COLUMNA SPECIES: 

In [ ]:
Especies= reemplazar1["Species "].unique()

In [ ]:
list(Especies)

In [ ]:
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ww](hite|HITE).*","White shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Tt](iger|IGER).*","Tiger shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ll](emon|EMON).*","Lemon shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Hh](ammerhead|AMMERHEAD).*","Hammerhead shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Bb](ull|ULL).*","Bull shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Bb](lue|LUE).*","Blue shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ss](ilvertip|ILVERTIP).*","Silvertip shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Nn](urse|URSE).*","Nurse shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ww](haler|HALER).*","Whaler shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Bb](lacktip|LACKTIP).*","Blacktip shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[MM](ako|AKO).*","Mako shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ss](and|AMD).*","Sand shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ww](obbegong|OBBEGONG).*","Wobbegong shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Gg](alapagos|ALAPAGOS).*","Galapagos shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Gg](rey|REY).*","Grey shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ll](eopard|EOPARD).*","Leopard shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Zz](ambesi|AMBESI).*","Zambesi shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Bb](lacktail|LACKTAIL).*","Blacktail shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Rr](ed|ED).*","Red shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Dd](usky|USKY).*","Dusky shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Rr](aggedtooth|AGGEDTOOTH).*","Raggedtooth shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ss](pinner|PINNER).*","Spinner shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Cc](ow|OW).*","Cow shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Pp](orbeagle|ORBEAGLE).*","Porbeagle shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace("[^.?!]*Bronze[^.?!]*","Bronze whaler shark",regex=True) #este no sale cuando filtro por "Bronze whaler shark". Ahora me desaparece Bronze directamente...
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Cc](aribbean|ARIBBEAN).*","Caribbean reef shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ss](and|AND).*","Sandbar shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ss](ilky|ILKY).*","Silky shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Zz](ambezi|AMBEZI).*","Zambezi shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ss](evengill|EVENGILL).*","Sevengill shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Cc](opper|OPPER).*","Copper shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Aa](ngel|NGEL)\s","Angel shark",regex=True) #lo he puesto diferente para que no pille "los ángeles" aún así parece que no ha funcionado.
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Ss](almon|ALMON).*","Salmon shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Gg](oblin|OBLIN).*","Goblin shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Tt](hresher|HRESHER).*","Thresher shark",regex=True)
reemplazar1["Species "]= reemplazar1["Species "].str.replace(".*[Dd](ogfish|OGFISH).*","Dogfish shark",regex=True)
reemplazar1["Species "]=reemplazar1["Species "].str.replace("[^.?!]*involvement[^.?!]*","Involvement not confirmed",regex=True)#no funciona

In [77]:
Fecha= reemplazar1["Date"].unique()
Fecha

array(['25-Jun-2018', '18-Jun-2018', '09-Jun-2018', ..., '1900-1905',
       '1883-1889', '1845-1853'], dtype=object)

In [78]:
list(Fecha)

['25-Jun-2018',
 '18-Jun-2018',
 '09-Jun-2018',
 '08-Jun-2018',
 '04-Jun-2018',
 '03-Jun-2018',
 '27-May-2018',
 '26-May-2018',
 '24-May-2018',
 '21-May-2018',
 '13-May-2018',
 'May 2018',
 '12-May-2018',
 '09-May-2018',
 'Reported 30-Apr-2018',
 '28-Apr-2018',
 '25-Apr-2018',
 '24-Apr-2018',
 '23-Apr-2018',
 '22-Apr-2018',
 '19-Apr-2018',
 '15-Apr-2018',
 '14-Apr-2018',
 'Reported 10-Apr-2018',
 '09-Apr-2018',
 '05-Apr-2018',
 '03-Apr-2018',
 '31-Mar-2018',
 '14-Mar-2018',
 '9-Mar-2018',
 '24-Feb-2018',
 '23-Feb-2018',
 '18-Feb-2018',
 '15-Feb-2018',
 '14-Feb-2018',
 '11-Feb-2018',
 '03-Feb-2018',
 '01-Feb-2018',
 '28-Jan-2018',
 '21-Jan-2018',
 '14-Jan-2018',
 '13-Jan-2018',
 '12-Jan-2018',
 '05-Jan-2018',
 '31-Dec-2017',
 '30-Dec-2017',
 '21-Dec-2017',
 '09-Dec-2017',
 '30-Nov-2017',
 'Reported 25-Nov-2017',
 '24-Nov-2017',
 '18-Nov-2017',
 'Reported 13-Nov-2017',
 '13-Nov-2017',
 '04-Nov-2017',
 'Reported 31-Oct-2017',
 '28-Oct-2017',
 '26-Oct-2017',
 '23-Oct-2017',
 '22-Oct-2017',

In [162]:
reemplazar1["Date_month"] = reemplazar1["Date"].apply(cf.mes) #estoy importando una función de visual

In [ ]:
#esta es la función que me he hecho para poder agrupar solo las fechas por mes en una nueva columna, pero no he sido capaz de importarla desde Visual

In [81]:
#def mes (x):
    #for i in x:
        #if "Jan" in x:
            #return "January"
        #elif "Feb" in x:
            #return "February"
        #elif "Mar" in x:
            #return "March"
        #elif "Apr" in x:
            #return "April"
        #elif "May" in x:
            #return "May"
        #elif "Jun" in x:
            #return "June"
        #elif "Jul" in x:
            #return "July"
        #elif "Aug" in x:
            #return "August"
        #elif "Sep" in x:
            #return "September"
        #elif "Oct" in x:
            #return "October"
        #elif "Nov" in x:
            #return "November"
        #elif "Dec" in x:
            #return "December"
        #else:
            #return "Other"

In [160]:
#reemplazar1["Date"].apply(mes)

0        June
1        June
2        June
3        June
4        June
        ...  
6297    Other
6298    Other
6299    Other
6300    Other
6301    Other
Name: Date, Length: 6301, dtype: object

Añado una nueva lista solo con los meses, a través de mi función "mes", que es lo que me interesa para comprobar mi hipótesis.

In [161]:
#reemplazar1["month"] = reemplazar1["Date"].apply(mes)

In [163]:
reemplazar1.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location Activity             Name Sex       Age  \
0     Oceanside, San Diego County    Other      Julie Wolfe    F       57   
1  St. Simon Island, Glynn County    Other  Adyson McNeely     F       11   
2                    Habush, Oahu    Other      John Denges    N       48   
3              Arrawarra Headland    Other             male    N  Unknown   
4                        La Ticla    Other   Gustavo Ramos     N  Unknown   

        Injury Fatal (Y/N)           Time         Species   \
0    No injury     Unknown          18h00      White shark   
1    No injury     Unknown  14h00  -15h00          Unknown   
2   Leg injury     Unknown          07h45          Unknown   
3    No injury     Unknown        Unknown        2 m shark   
4  Hand injury     Unknown        Unknown  Tiger shark, 3m   

           Investigator or Source month Date_month  
0                R. Collier, GSAF  June       Juny  
1  K.McMurray, TrackingSharks.com  June       Juny  
2  K.McMurray, TrackingSharks.com  June       Juny  
3                  B. Myatt, GSAF  June       Juny  
4                       A .Kipper  June       Juny

In [85]:
Type= reemplazar1["Type"].unique()
Type

array(['Boating', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster', 'Boat', 'Boatomg'], dtype=object)

In [86]:
reemplazar1.Type.replace({"Boat": "Boating", "Boatomg":"Boating","Questionable":"Invalid"}, inplace=True) 

In [87]:
reemplazar1["Type"].unique()

array(['Boating', 'Unprovoked', 'Invalid', 'Provoked', 'Sea Disaster'],
      dtype=object)

In [88]:
reemplazar1["Type"].value_counts()

Unprovoked      4594
Provoked         574
Invalid          553
Boating          341
Sea Disaster     239
Name: Type, dtype: int64

In [89]:
Area= reemplazar1["Area"].unique()
Area

array(['California', 'Georgia', 'Hawaii', 'New South Wales', 'Colima',
       'Pernambuco', 'Florida', 'Queensland', 'South Carolina',
       'Cornwall', 'Westerm Australia', 'Eastern Cape Province',
       'Hua Hin', 'Cocos Island', 'Western Australia',
       'Alifu Alifu Atoll', 'Western Cape Province', 'New Providence',
       'Alagoas', 'Unknown', 'Victoria', 'KwaZulu-Natal',
       'Boi Island, Victoria', 'Galapagos Islands', 'Fernando de Noronha',
       'Sepang', 'Holquin Province', 'Pamplemousses ', 'South Australia',
       'North Island', 'New York', 'Canary Islands', ' Upolu Island',
       'Shizuoka Prefecture', 'Texas', 'Castellón', 'Massachusetts',
       'Red Sea Protectorate', 'New Providence District',
       '40 miles off Grand Bahama Island', 'Ascension Island',
       'New Jersey', 'Majorca', 'Washington', 'Tabasco', 'Anjouan',
       'Ibiza Island', 'Marquesas', 'South Devon', 'New Providence ',
       'Sharjah, ', 'Baja California Sur', 'Saint-Leu', 'South Island

In [90]:
reemplazar1["Area"].value_counts() #entiendo que las áreas son las que son, no puedo limpiar esta columna

Florida                                 1037
New South Wales                          486
Unknown                                  455
Queensland                               311
Hawaii                                   297
                                        ... 
Ysabel Island                              1
 Lau Province                              1
South Coast, East New Britain              1
Between Southampton & Canary Islands       1
Moala Island                               1
Name: Area, Length: 826, dtype: int64

In [91]:
Pais= reemplazar1["Country"].unique()
Pais

array(['USA', 'AUSTRALIA', 'MEXICO', 'BRAZIL', 'ENGLAND', 'SOUTH AFRICA',
       'THAILAND', 'COSTA RICA', 'MALDIVES', 'BAHAMAS', 'NEW CALEDONIA',
       'ECUADOR', 'MALAYSIA', 'LIBYA', 'Unknown', 'CUBA', 'MAURITIUS',
       'NEW ZEALAND', 'SPAIN', 'SAMOA', 'SOLOMON ISLANDS', 'JAPAN',
       'EGYPT', 'ST HELENA, British overseas territory', 'COMOROS',
       'REUNION', 'FRENCH POLYNESIA', 'UNITED KINGDOM',
       'UNITED ARAB EMIRATES', 'PHILIPPINES', 'INDONESIA', 'CHINA',
       'COLUMBIA', 'CAPE VERDE', 'Fiji', 'DOMINICAN REPUBLIC',
       'CAYMAN ISLANDS', 'ARUBA', 'MOZAMBIQUE', 'FIJI', 'PUERTO RICO',
       'ITALY', 'ATLANTIC OCEAN', 'GREECE', 'ST. MARTIN', 'FRANCE',
       'PAPUA NEW GUINEA', 'TRINIDAD & TOBAGO', 'KIRIBATI', 'ISRAEL',
       'DIEGO GARCIA', 'TAIWAN', 'JAMAICA', 'PALESTINIAN TERRITORIES',
       'GUAM', 'SEYCHELLES', 'BELIZE', 'NIGERIA', 'TONGA', 'SCOTLAND',
       'CANADA', 'CROATIA', 'SAUDI ARABIA', 'CHILE', 'ANTIGUA', 'KENYA',
       'RUSSIA', 'TURKS & CAICOS', 

In [92]:
reemplazar1["Country"].value_counts() #entiendo que los países son los que son, no puedo limpiar esta columna

USA                      2228
AUSTRALIA                1338
SOUTH AFRICA              579
PAPUA NEW GUINEA          134
NEW ZEALAND               128
                         ... 
THE BALKANS                 1
NORTH ATLANTIC OCEAN        1
MAYOTTE                     1
GABON                       1
CEYLON (SRI LANKA)          1
Name: Country, Length: 213, dtype: int64

In [93]:
Nombre= reemplazar1["Name"].unique()
list(Nombre)

['Julie Wolfe',
 'Adyson\xa0McNeely ',
 'John Denges',
 'male',
 'Gustavo Ramos ',
 'Chris \x85',
 'Jose Ernesto da Silva ',
 'Cody High',
 'Trey de Boer',
 'Jei Turrell',
 'Max Berryman',
 'Melisa Brunning',
 'female',
 'Adam Murray',
 'Matthew Lee',
 'Rob Bruce',
 'Nathan Burch',
 'Ahmed Rasheed',
 'Dylan McWilliams',
 'Werner Danielsen',
 'Pablo de Melo',
 'Jason Longrass',
 'Alejandro Travaglini',
 'Bruce Rowan',
 'Josias Paz',
 'Unknown',
 'Shane McConnell',
 'Ross Spowart',
 'Luke Guy & Finn Bald',
 'Rob Crossland',
 'Lachie Brown',
 'Elton Polly',
 'Anna Shurapey ',
 'Brett Palmer',
 'Adam Hoare',
 'Sharna Babd',
 'Rosalida Souza',
 'Andrew Phipps Newman',
 'Lucas Martin',
 'Mathew Vickers',
 'Callum Stewart',
 'Ricardo Ferrari Bulhoes ',
 'Justin ',
 'Marjorie Mariano',
 'Natalie Jones',
 'Sairol  Harun',
 'Sarah Illig-Carroll',
 '__ Jimenez',
 'Rohina Bhandari',
 '31 migrants',
 'Grigor Azatian',
 'Kaia Anderson',
 'Timur Yunusov',
 'Charlie Fry',
 'Jesús Cabrera González ',
 

In [94]:
frecuencia_nombres=reemplazar1["Name"].value_counts(sort=False) #Para que la frecuencia esté indicada con el mismo orden que la lista de valores únicos tengo que incluir la función sort=False
frecuencia_nombres

Julie Wolfe           1
Adyson McNeely        1
John Denges           1
male                550
Gustavo Ramos         1
                   ... 
a young Scotsman      1
Mr. Masury            1
A "chiefess"          1
Ahmun                 1
Jules Patterson       1
Name: Name, Length: 5230, dtype: int64

In [45]:
dicc_nombres ={}
for Nombre, frecuencia_nombres in zip(Nombre, frecuencia_nombres):
    dicc_nombres[Nombre] = frecuencia_nombres

In [46]:
dicc_nombres

{'Julie Wolfe': 1,
 'Adyson\xa0McNeely ': 1,
 'John Denges': 1,
 'male': 550,
 'Gustavo Ramos ': 1,
 'Chris \x85': 1,
 'Jose Ernesto da Silva ': 1,
 'Cody High': 1,
 'Trey de Boer': 1,
 'Jei Turrell': 1,
 'Max Berryman': 1,
 'Melisa Brunning': 1,
 'female': 97,
 'Adam Murray': 1,
 'Matthew Lee': 1,
 'Rob Bruce': 1,
 'Nathan Burch': 1,
 'Ahmed Rasheed': 1,
 'Dylan McWilliams': 1,
 'Werner Danielsen': 1,
 'Pablo de Melo': 1,
 'Jason Longrass': 1,
 'Alejandro Travaglini': 1,
 'Bruce Rowan': 1,
 'Josias Paz': 1,
 'Unknown': 214,
 'Shane McConnell': 1,
 'Ross Spowart': 1,
 'Luke Guy & Finn Bald': 1,
 'Rob Crossland': 1,
 'Lachie Brown': 1,
 'Elton Polly': 1,
 'Anna Shurapey ': 1,
 'Brett Palmer': 1,
 'Adam Hoare': 1,
 'Sharna Babd': 1,
 'Rosalida Souza': 1,
 'Andrew Phipps Newman': 1,
 'Lucas Martin': 1,
 'Mathew Vickers': 2,
 'Callum Stewart': 1,
 'Ricardo Ferrari Bulhoes ': 1,
 'Justin ': 1,
 'Marjorie Mariano': 1,
 'Natalie Jones': 1,
 'Sairol  Harun': 1,
 'Sarah Illig-Carroll': 1,
 '__ 

In [47]:
nombres_unknown= dict(filter(lambda x: x[1] >3, dicc_nombres.items()))
nombres_unknown

{'male': 550,
 'female': 97,
 'Unknown': 214,
 'girl': 7,
 'child': 12,
 'boy': 23,
 '2 males': 16,
 'Unidentified': 6,
 'Anonymous': 10,
 'black male': 4,
 '2 fishermen': 4,
 'a soldier': 5,
 'boat': 14,
 'fisherman': 6,
 'a pearl diver': 6,
 'sailor': 10,
 'a sailor': 8,
 'males': 10,
 'a native': 5,
 'Arab boy': 4}

Quiero que todos estos valores se renombren como Unknown. Lo voy a hacer con la misma función usada anteriormente para la columna de Activity.

In [95]:
Daños= reemplazar1["Injury"].unique()

In [96]:
list(Daños)

['No injury to occupant, outrigger canoe and paddle damaged',
 'Minor injury to left thigh',
 'Injury to left lower leg from surfboard skeg',
 'Minor injury to lower leg',
 'Lacerations to leg & hand shark PROVOKED INCIDENT',
 'No injury, board bitten',
 'FATAL',
 'Minor injury to foot. PROVOKED INCIDENT',
 'Lower left leg bitten',
 'Minor injury to foot',
 'Minor bite to hand by captive shark. PROVOKED INCIDENT',
 'Injuries to lower right leg and foot',
 'Severe bite to right forearm',
 'Injured by teeth of a dead porbeagle shark he was tossing overboard.',
 'Shallow lacerations to finger PROVOKED INCIDENT',
 'Minor injury, marks on board',
 'No injury, surfboard damaged',
 'Minor injury to ankle',
 'No injury, shark bit scuba gear',
 'No injury, knocked off board by shark',
 'No injury',
 'No injury, shark struck his leg',
 '5-inch cut to hand',
 'No injury, shark bit hole in ski',
 'Lacerations to right lower leg',
 'Lacerations to foot and ankle',
 'Multiple severe injuries to arms

In [113]:
len(Daños)

3737

Voy a agrupar algunos daños que se repiten en valores únicos

In [110]:
reemplazar1["Injury"]= reemplazar1["Injury"].str.replace(".*[Nn](o|O)\s*[Ii](njury|NJURY).*","No injury",regex=True)

In [111]:
list(reemplazar1["Injury"].unique())

['No injury',
 'Minor injury to left thigh',
 'Injury to left lower leg from surfboard skeg',
 'Minor injury to lower leg',
 'Lacerations to leg & hand shark PROVOKED INCIDENT',
 'FATAL',
 'Minor injury to foot. PROVOKED INCIDENT',
 'Lower left leg bitten',
 'Minor injury to foot',
 'Minor bite to hand by captive shark. PROVOKED INCIDENT',
 'Injuries to lower right leg and foot',
 'Severe bite to right forearm',
 'Injured by teeth of a dead porbeagle shark he was tossing overboard.',
 'Shallow lacerations to finger PROVOKED INCIDENT',
 'Minor injury, marks on board',
 'Minor injury to ankle',
 '5-inch cut to hand',
 'Lacerations to right lower leg',
 'Lacerations to foot and ankle',
 'Multiple severe injuries to arms and leg, leg subsequently surgically amputated',
 'Laceration to right thigh',
 'Lacerations to legs',
 'Injury to ankle from marine animal trapped in weir PROVOKED INCIDENT. ',
 '2 puncture wounds to left foot, abrasion to right foot',
 'Lacerations to left knee & lower l

In [114]:
len(Daños) #¿Cómo puede ser que me salga la misma lontigud si se han agrupado

3737

In [139]:
reemplazar1["Injury"]= reemplazar1["Injury"].str.replace(".*[Hh](and|AND).*","Hand injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Mm](inor|INOR).*","Minor injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Ll](eg|LEG).*","Leg injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Ff](atal|ATAL).*","Fatal injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Ff](oot|OOT).*","Foot injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Sv](evere|EVERE).*","Severe injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Hh](ead|EAD).*","Head injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Ll](aceration|ACERATION).*","Laceration injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Aa](bdomen|BDOMEN).*","Abdomen injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Aa](brasion|BRASION).*","Abrasion injury",regex=True)
reemplazar1["Injury"]=reemplazar1["Injury"].str.replace(".*[Tt](oe|OE).*","Toe injury",regex=True)

In [140]:
list(reemplazar1["Injury"].unique())

['No injury',
 'Leg injury',
 'Hand injury',
 'Fatal injury',
 'Severe injury',
 'Injured by teeth of a dead porbeagle shark he was tossing overboard.',
 'Laceration injury',
 'Foot injury',
 'Injury to ankle from marine animal trapped in weir PROVOKED INCIDENT. ',
 'Minor injury',
 ' The shark & man simply collided; neither were injured',
 'Abdomen injury',
 'Heel bitten',
 'Puncture wounds to feet',
 'Puncture wounds to left shoulder',
 'Calf injured',
 'Abrasion injury',
 'Injury to ankle & feet',
 'Not injured by shark but board bitten',
 'Torso bitten',
 'Skull found in shark, a probable drowning & scavenging',
 'Arm injured by hook, not by a shark',
 'Fingernail pulled off ',
 'Right forearm severed ',
 'Thigh nipped PROVOKED INCIDENT',
 'Major injury to posterior thigh',
 'Calf bitten',
 'Right thigh injured ',
 'Multiple injuries to arm, wrist and torso',
 'Injury to right wrist & forearm',
 'No attack, shark made a threat display',
 'Buttock bitten',
 'Injuries to right calf a

In [97]:
Fatal= reemplazar1["Fatal (Y/N)"].unique()

In [98]:
list(Fatal)

['Unknown', 'Y', 'M', 'UNKNOWN', '2017', ' N', 'N ', 'y']

In [99]:
reemplazar1.replace({"M": "N", "UNKNOWN":"Unknown","2017":"Unknown"," N": "N","N ":"N","y":"Y"}, inplace=True) 

In [101]:
reemplazar1["Fatal (Y/N)"].unique()

array(['Unknown', 'Y', 'N'], dtype=object)

In [102]:
Time= reemplazar1["Time"].unique()

In [103]:
Time #en caso de que mi hipótesis necesite de los datos de Time para comprobarla, crearía una columna con los rangos horarios que necesite, mediante una función

array(['18h00', '14h00  -15h00', '07h45', 'Unknown', 'Late afternoon',
       '17h00', '14h00', 'Morning', '15h00', '08h15', '11h00', '10h30',
       '10h40', '16h50', '07h00', '09h30', 'Afternoon', '21h50', '09h40',
       '08h00', '17h35', '15h30', '07h30', '19h00, Dusk', 'Night',
       '16h00', '15h01', '12h00', '13h45', '23h30', '09h00', '14h30',
       '18h30', '12h30', '16h30', '18h45', '06h00', '10h00', '10h44',
       '13h19', 'Midday', '13h30', '10h45', '11h20', '11h45', '19h30',
       '08h30', '15h45', 'Shortly before 12h00', '17h34', '17h10',
       '11h15', '08h50', '17h45', '13h00', '10h20', '13h20', '02h00',
       '09h50', '11h30', '17h30', '9h00', '10h43', 'After noon', '15h15',
       '15h40', '19h05', '1300', '14h30 / 15h30', '22h00', '16h20',
       '14h34', '15h25', '14h55', '17h46', 'Morning ', '15h49', '19h00',
       'Midnight', '09h30 / 10h00', '10h15', '18h15', '04h00', '14h50',
       '13h50', '19h20', '10h25', '10h45-11h15', '16h45', '15h52',
       '06h15'

In [104]:
Investigador= reemplazar1["Investigator or Source"].unique()

In [105]:
list(Investigador) #no me da ninguna información relevante que se haya investigado por una u otra persona

['R. Collier, GSAF',
 'K.McMurray, TrackingSharks.com',
 'B. Myatt, GSAF',
 'A .Kipper',
 'Daily Telegraph, 6/4/2018',
 'Diario de Pernambuco, 6/4/2018',
 'K. McMurray, TrackingSharks.com',
 'K. McMurray, Tracking Sharks.com',
 'ABC.net.au ,05/24/2018',
 'C. Creswell, GSAF',
 'C. Creswell, GSAF & K. McMurray TrackingSharks.com',
 'Perth Now, 6/30/2018',
 'B.Jacob, Dispatch Live, 5/12/2018',
 'Costa Rica Star. 4/28/2018',
 'B.Myatt, GSAF',
 'Africa News Agency, 4/22/2018',
 'The Garden Island, 4/19/2018',
 'Straits Times, 4/16/2018',
 'Globo, 4/16/2018',
 'Tribune242,',
 'Les Nouvelles Caledoniennes, 4/10/2018',
 'Khon2, 3/31/2018',
 'International Business Times, 3/1/2018',
 'Sunshine Coast Daily, 2/15/2018',
 'TrackingSharks.com',
 'The Standard (UK), 2/6/2018',
 'The West Australian, 2/2/2018',
 'ABC.net.au, 01/28/2018',
 'Perth Now, 1/6/2018',
 'J. Howard, Surfling Now, 1/2/2018',
 'R. Collier',
 'Daily Star, 12/11/2017',
 'R. Arauz',
 'TG Com 24, 11/25/2017',
 'Florida Today, 11/21

In [106]:
reemplazar1.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex       Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F       57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F       11   
2                    Habush, Oahu      Surfing      John Denges    N       48   
3              Arrawarra Headland      Surfing             male    N  Unknown   
4                        La Ticla  Free diving   Gustavo Ramos     N  Unknown   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...     Unknown   
1                         Minor injury to left thigh     Unknown   
2       Injury to left lower leg from surfboard skeg     Unknown   
3                          Minor injury to lower leg     Unknown   
4  Lacerations to leg & hand shark PROVOKED INCIDENT     Unknown   

            Time         Species           Investigator or Source month  
0          18h00      White shark                R. Collier, GSAF  June  
1  14h00  -15h00          Unknown  K.McMurray, TrackingSharks.com  June  
2          07h45          Unknown  K.McMurray, TrackingSharks.com  June  
3        Unknown        2 m shark                  B. Myatt, GSAF  June  
4        Unknown  Tiger shark, 3m                       A .Kipper  June

Voy a crearme un dataframe solo con las columnas que necesito para comprobar mis hipótesis

In [148]:
reemplazar_hipotesis=reemplazar1[["Sex ", "Activity", "Date", "Area", "Species ", "Country","month"]]

In [149]:
reemplazar_hipotesis.head()

Sex      Activity         Date             Area         Species     Country  \
0    F     Paddling  25-Jun-2018       California      White shark        USA   
1    F     Standing  18-Jun-2018          Georgia          Unknown        USA   
2    N      Surfing  09-Jun-2018           Hawaii          Unknown        USA   
3    N      Surfing  08-Jun-2018  New South Wales        2 m shark  AUSTRALIA   
4    N  Free diving  04-Jun-2018           Colima  Tiger shark, 3m     MEXICO   

  month  
0  June  
1  June  
2  June  
3  June  
4  June

Para comprobar mis hipótesis necesitaré agrupar las columnas que quiero comparar

In [142]:
reemplazar1.groupby(["Sex ", "Activity"]).count().head(15)

Date  Year  Type  \
Sex  Activity                                                               
F    "Riding waves on a board"                              1     1     1   
     16' catamaran capsized previous night, occupant...     1     1     1   
     17 Maldivians adrift in open boat for 31 days          1     1     1   
     A father bathing his smallest daughter when the...     1     1     1   
     Air Disaster                                           1     1     1   
     Attempting to attract dolphins                         1     1     1   
     Attempting to rescue a shark                           1     1     1   
     Attempting to swim across the Atlantic Ocean           1     1     1   
     Bathing                                               16    16    16   
     Bathing in 5' of water                                 1     1     1   
     Bathing in river                                       1     1     1   
     Bathing or body surfing                                1     1     1   
     Bathing or washing                                     1     1     1   
     Bathing with her mother                                1     1     1   
     Batin                                                  1     1     1   

                                                         Country  Area  \
Sex  Activity                                                            
F    "Riding waves on a board"                                 1     1   
     16' catamaran capsized previous night, occupant...        1     1   
     17 Maldivians adrift in open boat for 31 days             1     1   
     A father bathing his smallest daughter when the...        1     1   
     Air Disaster                                              1     1   
     Attempting to attract dolphins                            1     1   
     Attempting to rescue a shark                              1     1   
     Attempting to swim across the Atlantic Ocean              1     1   
     Bathing                                                  16    16   
     Bathing in 5' of water                                    1     1   
     Bathing in river                                          1     1   
     Bathing or body surfing                                   1     1   
     Bathing or washing                                        1     1   
     Bathing with her mother                                   1     1   
     Batin                                                     1     1   

                                                         Location  Name  Age  \
Sex  Activity                                                                  
F    "Riding waves on a board"                                  1     1    1   
     16' catamaran capsized previous night, occupant...         1     1    1   
     17 Maldivians adrift in open boat for 31 days              1     1    1   
     A father bathing his smallest daughter when the...         1     1    1   
     Air Disaster                                               1     1    1   
     Attempting to attract dolphins                             1     1    1   
     Attempting to rescue a shark                               1     1    1   
     Attempting to swim across the Atlantic Ocean               1     1    1   
     Bathing                                                   16    16   16   
     Bathing in 5' of water                                     1     1    1   
     Bathing in river                                           1     1    1   
     Bathing or body surfing                                    1     1    1   
     Bathing or washing                                         1     1    1   
     Bathing with her mother                                    1     1    1   
     Batin                                                      1     1    1   

                                                         Injury  Fatal (Y/N)  \
Sex  Activity                                  